merlin/models/tf/blocks/core/index.py

```diff
@@ -0,0 +1,229 @@
+        output_dim: int
+            The output dimension of `block`.
+        data: merlin.io.Dataset
+            Dataset containing raw item features.
+        k: int
+            Number of top candidates to retrieve.
+            Defaults to 20
+        id_column: Optional[str]
+            The candidates ids column name.
+            Note, this will be inferred automatically if the block contains
+            a schema with an item-id Tag.
+        """
+        return super().from_block(block=block, data=data, id_column=id_column, k=k, **kwargs)
+
+    def call(self, inputs: tf.Tensor, k=None, **kwargs) -> Union[tf.Tensor, tf.Tensor]:
+        """
+        Compute Top-k scores and related indices from query inputs
+
+        Parameters:
+        ----------
+        inputs: tf.Tensor
+            Tensor of pre-computed query embeddings.
+        k: int
+            Number of top candidates to retrieve
+            Defaults to constructor `_k` parameter.
+        Returns
+        -------
+        top_scores, top_indices: tf.Tensor, tf.Tensor
+            2D Tensors with the scores for the top-k candidates and related ids.
+        """
+        k = k if k is not None else self._k
+        scores = tf.matmul(inputs, self.values, transpose_b=True)
+        top_scores, top_indices = tf.math.top_k(scores, k=k)
+        top_indices = tf.gather(self.ids, top_indices)
+
+        return top_scores, top_indices
+
+    def call_outputs(
+        self, outputs: PredictionOutput, training=False, **kwargs
+    ) -> "PredictionOutput":
+        """
+        Retrieve top-k negative scores for evaluation.
+
+        Parameters
+        ----------
+        predictions: tf.Tensor
+            Tensor of pre-computed positive scores.
+            If`training=True`, the first column of predictions is expected
+            to be positive scores and the remaining sampled negatives are ignored.
+
+        Returns
+        -------
+        targets, predictions: tf.Tensor, tf.Tensor
+            2D Tensors with the one-hot representation of true targets and
+            the scores for the top-k implicit negatives.
+        """
+        targets, predictions = outputs.targets, outputs.predictions
+        queries = self.context["query"]
+        top_scores, _ = self(queries, k=self._k)
+        predictions = tf.expand_dims(predictions[:, 0], -1)
+        predictions = tf.concat([predictions, top_scores], axis=-1)
+        # Positives in the first column and negatives in the subsequent columns
+        targets = tf.concat(
+            [
+                tf.ones([tf.shape(predictions)[0], 1]),
+                tf.zeros([tf.shape(predictions)[0], self._k]),
+            ],
+            axis=1,
+        )
+        return PredictionOutput(predictions, targets)
+
+    def compute_output_shape(self, input_shape):
+        batch_size = input_shape[0]
+        return tf.TensorShape((batch_size, self._k)), tf.TensorShape((batch_size, self._k))
```

Comment:
Shouldn't the tensor shapes be `(batch size, k+1)` ?

Reply:
The current implementation supposes that top-k includes the positive scores so we are retrieving `k-1` negatives.